File: files/400-499/403/CH-403.R
library(tidyverse)
library(readxl)
library(memoise)

path <- "400-499/403/CH-403 Number Puzzles - Happy Numbers.xlsx"
input <- read_excel(path, range = "B3:B10")
test <- read_excel(path, range = "D3:D10")

is_happy <- function(x) {
  step <- function(n) sum((as.integer(strsplit(as.character(n), "")[[1]]))^2)
  detect_index(
    accumulate(rep(NA, 100), ~ step(ifelse(.y == 1, x, .x)), .init = x),
    ~ .x == 1
  ) >
    0
}

result = input %>%
  mutate(is_happy = map_lgl(Number, is_happy))

File: files/400-499/403/CH-403.py
import pandas as pd

path = "400-499/403/CH-403 Number Puzzles - Happy Numbers.xlsx"
input = pd.read_excel(path, usecols="B", nrows=8, skiprows=2)
test = pd.read_excel(path, usecols="D", nrows=8, skiprows=2)

def is_happy(n):
    seen = set()
    while n != 1 and n not in seen:
        seen.add(n)
        n = sum(int(d)**2 for d in str(n))
    return n == 1

input['Is Happy?'] = input['Number'].apply(is_happy)

print(input['Is Happy?'].equals(test['Is Happy?']))